core/trino-main/src/main/java/io/trino/sql/planner/rowpattern/IrPatternAlternationOptimizer.java

```diff
@@ -0,0 +1,145 @@
+    public static IrRowPattern optimize(IrRowPattern node)
+    {
+        return new IrPatternAlternationOptimizer().process(node);
+    }
+
+    @Override
+    protected IrRowPattern visitIrRowPattern(IrRowPattern node, Void context)
+    {
+        throw new UnsupportedOperationException("unsupported node type: " + node.getClass().getName());
+    }
+
+    @Override
+    protected IrRowPattern visitIrLabel(IrLabel node, Void context)
+    {
+        return node;
+    }
+
+    @Override
+    protected IrRowPattern visitIrAnchorPattern(IrAnchorPattern node, Void context)
+    {
+        return node;
+    }
+
+    @Override
+    protected IrRowPattern visitIrEmptyPattern(IrEmptyPattern node, Void context)
+    {
+        return node;
+    }
+
+    @Override
+    protected IrRowPattern visitIrExcludedPattern(IrExcludedPattern node, Void context)
+    {
+        IrRowPattern child = process(node.getPattern());
+
+        return new IrExcludedPattern(child);
+    }
+
+    @Override
+    protected IrRowPattern visitIrPatternAlternation(IrPatternAlternation node, Void context)
+    {
+        List<IrRowPattern> children = node.getPatterns().stream()
+                .map(this::process)
+                .collect(toImmutableList());
+
+        List<IrRowPattern> emptyChildren = children.stream()
+                .filter(IrEmptyPattern.class::isInstance)
+                .collect(toImmutableList());
+        if (emptyChildren.isEmpty()) {
+            return new IrPatternAlternation(children);
+        }
+        checkState(emptyChildren.size() == 1 && children.size() > 1, "run IrRowPatternFlattener first to remove redundant empty pattern");
+
+        IrRowPattern emptyChild = getOnlyElement(emptyChildren);
+        int emptyChildIndex = children.indexOf(emptyChild);
+
+        // remove the empty child:
+        // (() | A) -> A??
+        // (() | A | B) -> (A?? | B)
+        if (emptyChildIndex == 0) {
+            IrRowPattern child = new IrQuantifiedPattern(children.get(1), zeroOrOne(false));
+            if (children.size() == 2) {
+                return child;
+            }
+            ImmutableList.Builder<IrRowPattern> builder = ImmutableList.<IrRowPattern>builder()
+                    .add(child);
+            for (int i = 2; i < children.size(); i++) {
+                builder.add(children.get(i));
+            }
+            return new IrPatternAlternation(builder.build());
+        }
+        // (A | ()) -> A?
+        // (A | B | () | C) -> (A | B? | C)
+        ImmutableList.Builder<IrRowPattern> builder = ImmutableList.builder();
+        for (int i = 0; i < emptyChildIndex - 1; i++) {
+            builder.add(children.get(i));
+        }
+        builder.add(new IrQuantifiedPattern(children.get(emptyChildIndex - 1), zeroOrOne(true)));
+        for (int i = emptyChildIndex + 1; i < children.size(); i++) {
+            builder.add(children.get(i));
+        }
+        children = builder.build();
+        if (children.size() == 1) {
```

Comment:
```java
        children = ImmutableList.<IrRowPattern>builder()
                .addAll(children.subList(0, emptyChildIndex - 1))
                .add(new IrQuantifiedPattern(children.get(emptyChildIndex - 1), zeroOrOne(true)))
                .addAll(children.subList(emptyChildIndex + 1, children.size()))
                .build();
```

but make sure you don't run into IndexOutOfBounds issues (vs the `for` loops, which would just skip that part)